w3af/plugins/audit/hpp.py

```diff
@@ -0,0 +1,263 @@
+                    result += [p, ref]
+
+    return result
+
+
+class hpp(AuditPlugin):
+    """
+    Identify client-side http parameter pollution vulnerabilities.
+    """
+
+    def __init__(self):
+        AuditPlugin.__init__(self)
+        self.parser = etree.HTMLParser(remove_blank_text=True, remove_comments=True)
+
+    def audit(self, freq, orig_response):
+        """
+        Tests an URL for HPP vulnerabilities.
+        :author: Dmitry Roshchin (nixwizard@gmail.com)
+        """
+
+        if not orig_response.is_text_or_html():
+            return
+
+        if not orig_response.get_code() == 200:
+            return
+
+        method = freq.get_method()
+
+        if method == 'GET':
+            qs = freq.get_querystring()
+
+        precedence = None
+        #finding reflected params if any
+        for param in qs:
+            value = qs[param][0]
+            #checking if the value is number or string so that our
+            #reflection test request would be more legitimate
+            if value.isdigit():
+                newval = unicode(int(value)+1)
+            else:
+                newval = REFLECT_TOKEN
+
+            if precedence:
+                continue
+
+            mutants = create_mutants(freq, [newval],
+                                     fuzzable_param_list=[param],
+                                     mutant_tuple=(QSMutant,))
+            mu = mutants[0]
+            #doing reflection test for the param
+            rtest_resp = self._uri_opener.send_mutant(mu)
+            if not rtest_resp.get_code() == 200:
+                #no need to seek reflection if it is not OK
+                continue
+
+            if orig_response.body == rtest_resp.body:
+                #parameter param is not being reflected going for the next
+                continue
+
+            #Now we are doing precedence test
+            u = str(mu.get_uri())
+            #this hack is to make w3af send request with unencoded qs
+            #because as long as we are trying to use qs it urlencodes the
+            #second instance of param with new value
+            ptest_url = URL(u.replace(newval,
+                                      value + '&' + param + '=' + newval))
+            ptest_resp = self._uri_opener.GET(ptest_url)
+            ptest_resp_code = ptest_resp.get_code()
+
+            if not ptest_resp_code == 200:
+                om.out.debug(u"Precedence check request to the uri %s "
+                             u"failed! "
+                             u"Got response code %s in response id %s\n"
+                             u"Unable to continue HPP detection. "
+                             u"Resource might be protected by a WAF,"
+                             % (ptest_url, ptest_resp_code, ptest_resp.id))
+                #no need to continue with the param if it is not 200 OK
+                continue
+
+            ptest_body = ptest_resp.body
+
+            if fuzzy_equal(orig_response.body,
```

Comment:
Do we really need the precedence check for client-side? Isn't this too complex? Why not just set each input parameter to `value&param=3` , then parse the response for links, and check if the links have the extra-unexpected `param` parameter?

I'm getting really confused with all this...